client/src/partials/stock/StockForm.service.js

```diff
@@ -0,0 +1,165 @@
+angular.module('bhima.services')
+.service('StockFormService', StockFormService);
+
+StockFormService.$inject = [
+  'Store', 'AppCache', 'SessionService',
+  '$timeout',
+];
+
+/**
+ * @class StockFormService
+ *
+ * @description
+ * Provides utilities and functions for handling inventories and lots
+ *
+ * @todo - implement the cache feature
+ */
+function StockFormService(Store, AppCache, Session, $timeout) {
+  /**
+   * @constructor
+   */
+  function StockForm(cacheKey) {
+    if (!cacheKey) {
+      throw new Error('StockForm expected a cacheKey, but it was not provided.');
+    }
+
+    this.cache = AppCache(cacheKey);
+    this.details = {};
+    this.store = new Store({ identifier: 'index', data: [] });
+  }
+
+  /**
+   * @method setup
+   *
+   * @description
+   * This function initializes the journal voucher form with data.  By default,
+   * two lines are always present in the form.
+   */
+  StockForm.prototype.setup = function setup() {
+    var self = this;
+    self.details = {};
+    self.details.date = new Date();
+    self.details.user_id = Session.user.id;
+  };
+
+  /**
+   * @method addItems
+   *
+   * @description
+   * Adds an item to the grid.
+   *
+   * @param {Number} n - the number of items to add to the grid
+   */
+  StockForm.prototype.addItems = function addItems(n) {
+    var elt;
+    var idx;
+    var self = this;
+    while (n--) {
+      elt = new StockItem();
+      idx = self.store.data.length;
+      elt.index = idx ? idx + 1 : idx;
```

Comment:
`idx` is 0 when there is no data in the `StockForm.store.data`, so `elt` is the new element to add, it must have index 0 as the first element of `StockForm.store.data`.